Document how users can use pytest instead of kedro test (#1879)
kedro-org/kedro: docs/source/development/automated_testing.md

As a project grows larger, new code will increasingly rely on existing code. As these interdependencies grow, making changes in one part of the code base can unexpectedly break the intended functionality in another part.

The major disadvantage of manual testing is that it is time-consuming. Manual tests are usually run once, directly after new functionality has been added. It is impractical to repeat the manual tests for the entire existing code base each time a change is made, which means this strategy often misses breaking changes. 

The solution to this problem is automated testing, which allows many tests across the whole code base to be run in seconds, every time a new feature is added or an old one is changed. In this way, breaking changes can be discovered during development rather than in production.

## Set up automated testing with `pytest`

There are many testing frameworks available for Python. One of the most popular is [`pytest`](https://docs.pytest.org/). `pytest` is often used in Python projects for its short, readable tests and powerful set of features.

Let's look at how you can start working with `pytest` in your Kedro project.

### Install `pytest`

> Install `pytest` as you would install other packages with `pip`, making sure your project's virtual environment is active.

Author: Great idea, I've added this in [965c6fe](https://github.com/kedro-org/kedro/pull/1879/commits/965c6fe777c0a640e4ee3014dd5c75df800966d9).

```bash
pip install pytest
```

### Create a `/tests` directory

Now that `pytest` is installed, you will need a place to put your tests. Create a `/tests` folder in the `/src` directory of your project.

```bash
mkdir <project_root>/src/tests
```

### Test directory structure

> The subdirectories in your project's `/tests` directory should mirror the directory structure of your project's `/src/<project_name>` directory. All files in the `/tests` folder should be named `test_<file_being_tested>.py`. For instance:

Reviewer: ```suggestion
The subdirectories in your project's `/tests` directory should mirror the directory structure of your project's `/src/<package_name>` directory. All files in the `/tests` folder should be named `test_<file_being_tested>.py`. For instance:
```

We should stick with this naming convention.  https://github.com/kedro-org/kedro/pull/1753

Author: You're right about this. I changed three instances of <project_name> to <package_name> in [6a6a278](https://github.com/kedro-org/kedro/pull/1879/commits/6a6a2788d43e93c554e0e2381636df0ff1d366d7).

│           │   ...
```

### Create an example test

Now that you have a place to put your tests, you can create an example test in the new file `/src/tests/test_run.py`. The example test simply checks that the project_path attribute of a specially-defined `KedroContext` object has been correctly set.

```
import pytest
from kedro.config import ConfigLoader
from kedro.framework.context import KedroContext
from kedro.framework.hooks import _create_hook_manager

@pytest.fixture
def config_loader():
    return ConfigLoader(conf_source=str(Path.cwd()))

@pytest.fixture
def project_context(config_loader):
    return KedroContext(
        package_name=<package_name>,
        project_path=Path.cwd(),
        config_loader=config_loader,
        hook_manager=_create_hook_manager(),
    )

class TestProjectContext:
    def test_project_path(self, project_context):
        assert project_context.project_path == Path.cwd()
```

This test is redundant, but it introduces a few of `pytest`'s core features and demonstrates the layout of a test file:
- [Fixtures](https://docs.pytest.org/en/7.1.x/explanation/fixtures.html#about-fixtures) are used to define resources used in tests.

> - Tests are implemented in methods or functions beginning with `test_` and classes beginning with `Test`.

Author: Yep! Added in [ceecd5f](https://github.com/kedro-org/kedro/pull/1879/commits/ceecd5f2efd16997a2553a961174e9faccd57e7e)

- The `assert` statement is used to compare the result of the test with an expected value.

> You can read more about the basics of using `pytest` [here](https://docs.pytest.org/en/7.1.x/getting-started.html). For help writing your own tests and using all of the features of `pytest`, see the [docs](https://docs.pytest.org/).

Reviewer 2: ```suggestion
You can read more about the [basics of using `pytest` on the getting started page] (https://docs.pytest.org/en/7.1.x/getting-started.html). For help writing your own tests and using all of the features of `pytest`, see the [pytest docs](https://docs.pytest.org/).
```

Making the links a bit more descriptive for accessibility: https://github.com/kedro-org/kedro/issues/1281

Author: You're completely right about this. I fixed this in multiple places in [0a5795c](https://github.com/kedro-org/kedro/pull/1879/commits/0a5795c4e800d6f8a63a538470564d562707126d).

### Run your tests

> To run your tests, run `pytest` from within your project's root directory or from the `/src` directory.

Reviewer: Maybe we don't need need to mention the `src/`. In theory, `pytest` would still work, but the following section we also mention `pyproject.toml`, and it won't be picked up if it's running from `src/`

```bash
cd <project_root>
pytest
```

If you created the example test in the previous section, you should see the following output in your shell.

```
============================= test session starts ==============================
...
collected 1 item

src/tests/test_run.py .                                                  [100%]

============================== 1 passed in 0.38s ===============================
```

This output indicates that one test ran successfully in the file `src/tests/test_run.py`.

## Add test coverage reports with `pytest-cov`

It can be useful to see how much of your project is covered by tests. For this, you can install and configure the [`pytest-cov`](https://pypi.org/project/pytest-cov/) plugin for `pytest`, which is based on the popular [`coverage.py`](https://coverage.readthedocs.io/) library.

### Install `pytest-cov`

> Install `pytest` as you would install other packages with pip, making sure your project's virtual environment is active.

Reviewer 2: Could link to the virtual env setup docs again here.

Author: Added in [13586b6](https://github.com/kedro-org/kedro/pull/1879/commits/13586b61d8d9c81d5196e0358ee010ccb4e710e7).

--------------------------------------------------------------------------------------
src/spaceflights/__init__.py                                 1      1     0%   4
src/spaceflights/__main__.py                                30     30     0%   4-47
src/spaceflights/pipeline_registry.py                        7      7     0%   2-16
src/spaceflights/pipelines/__init__.py                       0      0   100%
src/spaceflights/pipelines/data_processing/__init__.py       1      1     0%   3
src/spaceflights/pipelines/data_processing/nodes.py         25     25     0%   1-67
src/spaceflights/pipelines/data_processing/pipeline.py       5      5     0%   1-8
src/spaceflights/pipelines/data_science/__init__.py          1      1     0%   3
src/spaceflights/pipelines/data_science/nodes.py            20     20     0%   1-55
src/spaceflights/pipelines/data_science/pipeline.py          8      8     0%   1-40
src/spaceflights/settings.py                                 0      0   100%
--------------------------------------------------------------------------------------
TOTAL                                                       98     98     0%
```

> This is the simplest report that `coverage.py` (via `pytest-cov`) will produce. It gives an overview of how many of the executable statements in each project file are covered by tests. For a full set of features offered by `coverage.py`, see the [docs](https://coverage.readthedocs.io/).

Reviewer 2: ```suggestion
This is the simplest report that `coverage.py` (via `pytest-cov`) will produce. It gives an overview of how many of the executable statements in each project file are covered by tests. For a full set of features offered by `coverage.py`, see the [coverage docs](https://coverage.readthedocs.io/).
```

Author: Fixed in [8650a0b](https://github.com/kedro-org/kedro/pull/1879/commits/8650a0bd5726382b78f1eeaa7439d121ac297ae3).
